docs/style/CODING_STYLE_GUIDE.md

```diff
@@ -0,0 +1,172 @@
+
+[restyle_link]:
+    https://github.com/project-chip/connectedhomeip/blob/master/.restyled.yaml
+[isort_link]:
+    https://github.com/project-chip/connectedhomeip/blob/master/.isort.cfg
+[ruff_link]:
+    https://github.com/project-chip/connectedhomeip/blob/master/ruff.toml
+
+All pull requests run formatting checks using these tools before merge is
+allowed. Generated code is not run through restyle.
+
+### C++
+
+#### Use C++ _cstdint_ for Plain Old Data Types
+
+Standard, scalar data types defined in _cstdint_ should be used for basic signed
+and unsigned integer types, especially when size and serialization to
+non-volatile storage or across a network is concerned.
+
+Examples of these are: `uint8_t`, `int8_t`, etc.
+
+#### Avoid top-level `using namespace` Statements in Headers
+
+By doing this, you are effectively forcing every other module that includes the
+header to also be using the namespace. This causes namespace pollution and
+generally defeats the purposes of namespaces. Fully-qualified symbols or
+namespace blocks should be used instead.
+
+#### Classes / objects not exposed in a header should be in an anonymous namespace
+
+If a cpp class defines a class or instantiates a static object, it should be
+enclosed in an anonymous namespace.
+
+```
+namespace {
+  // CPP internal defines go here
+}  // namespace
+```
+
+#### Singleton use
+
+The decision to use a singleton class should be considered with care. Do not
+default to using a singleton for ease of writing code.
+
+If the class truly should be a singleton (ex. if it is controlling access to a
+hardware resource)
+
+-   The standard function name for accessing an SDK singleton is GetInstance().
+-   Singleton classes should delete copy and move constructors
+
+#### Avoid Heap-based Resource Allocation and auto-resizing std containers
+
+Heap-based resource allocation should be avoided in the core SDK for common code
+that may run on constrained embedded devices. This includes any container
+element in std that automatically re-sizes itself at runtime (ex. vector, string
+etc.) as these re-size operations are often large and can lead to memory
+exhaustion and fragmentation on embedded systems.
+
+Heap-based allocation is allowed for controller code and is at the discretion of
+platform vendors for platform-specific code.
+
+##### Alternatives
+
+In either case, recommended resource allocation alternatives are:
+
+-   In-place allocation and initialization
+-   Pool-based allocators
+-   Platform-defined and -assigned allocators
+
+[CHIPMem.h](https://github.com/project-chip/connectedhomeip/blob/master/src/lib/support/CHIPMem.h)
+provides support for platform defined allocators.
+
+[Pool.h](https://github.com/project-chip/connectedhomeip/blob/master/src/lib/support/Pool.h)
+is the Matter SDK pool allocator implementation.
+
+#### Prefer CopySpanToMutableSpan over memcpy when using spans
```

Comment:
Should we also add a section for: Use Span/ByteSpan instead of C-style pointers for buffers ?

Reply:
Tracking suggestions for new content in a separate issue: https://github.com/project-chip/connectedhomeip/issues/36544